cai_causal_graph/identify_utils.py

```diff
@@ -92,30 +122,138 @@ def _identify_confounders_no_checks_no_descendant_pruning(clean_graph: CausalGra
 
-    # Coerce NodeLike to identifier. We already know they are NodeLike as node_exists does this.
-    node_1_id = Node.identifier_from(node_1)
-    node_2_id = Node.identifier_from(node_2)
-
-    # Ensure node_1 != node_2
-    if node_1_id == node_2_id or node_1 == node_2:
-        raise ValueError('node_1 and node_2 cannot be equal. Please provide different nodes / node identifiers.')
-
-    # Remove all descendants of node 1 and node 2 as they won't be necessary in the search for confounders.
-    # This is to reduce memory overhead as we pass the graph throughout recursive calls.
+    # Remove all descendants of node 1 and node 2 as they will not be necessary in the search for confounders
+    # This is to reduce memory overhead as we pass the graph throughout recursive calls
     pruned_graph = graph.copy()
-    all_descendants = pruned_graph.get_descendants(node_1).union(pruned_graph.get_descendants(node_2))
-    # Remove node 1 and node 2 identifiers if they are in descendants. Get descendants returns str so must work with id.
-    all_descendants -= {node_1_id, node_2_id}
+    all_descendants = pruned_graph.get_descendants(node_1_id).union(pruned_graph.get_descendants(node_2_id))
+    all_descendants -= {node_1_id, node_2_id}  # do not want to remove the nodes themselves
     for descendant in all_descendants:
         pruned_graph.remove_node(descendant)
 
     return _identify_confounders_no_checks_no_descendant_pruning(pruned_graph, node_1_id, node_2_id)
+
+
+def identify_instruments(graph: CausalGraph, source: NodeLike, destination: NodeLike) -> List[str]:
+    """
+    Identify all instrumental variables for the causal effect of `source` and `destination` in the provided `graph`.
+
+    An instrumental variable for the causal effect of `source` on `destination` satisfies the following criteria:
+        1. There is a causal effect between the `instrument` and the `source`.
+        2. The `instrument` has a causal effect on the `destination` _only_ through the `source`.
+        3. There is no confounding between the `instrument` and the `destination`.
+
+    Example:
+        >>> from typing import List
+        >>> from cai_causal_graph import CausalGraph
+        >>> from cai_causal_graph.identify_utils import identify_instruments
+        >>>
+        >>> # define a causal graph
+        >>> cg = CausalGraph()
+        >>> cg.add_edge('z', 'x')
+        >>> cg.add_edge('u', 'x')
+        >>> cg.add_edge('u', 'y')
+        >>> cg.add_edge('x', 'y')
+        >>>
+        >>> # find the instruments between 'x' and 'y'; output: ['z']
+        >>> instrumental_variables: List[str] = identify_instruments(cg, source='x', destination='y')
+
+    :param graph: The causal graph given by a `cai_causal_graph.causal_graph.CausalGraph` instance. This must be a DAG,
+        i.e. it must only contain directed edges and be acyclic, otherwise a `TypeError` is raised.
+    :param source: The source variable or its identifier.
```

Comment:
The source node or its identifier. Same for destination below. This matches language in confounder docstring too.